README.md
## What is the ND PHP Framework?

 ND PHP Framework is a secure, fully automated framework capable of delivering full featured web applications and respective REST API in a matter of minutes. It's licensed under [GPLv3](http://www.gnu.org/licenses/gpl-3.0.html).

## How does it work?

 Based on a drag and drop builder (the IDE Builder), you'll be able to create and visualize the application model. After the model is created, hit the Deploy button and the web application will be created, ready to be used.

## Do I need to code anything at all?

 No, if the base building blocks and features of the framework already fit your needs. At this moment, this is unlikely to happen unless you only want to rely on the REST API. If not, then you'll need to extend the features by providing further parameterizations and/or code development.

## How to install it?

 Download the current framework version and follow the instructions in the [INSTALL.txt](https://raw.githubusercontent.com/ucodev/nd-php/master/INSTALL.txt) file.

## Does it run in any version of PHP?

 It requires PHP version 5.6.0 or greater. It also runs under PHP 7.

## Do I need to install a lot of dependencies?

 Probably not. Probably your GNU/Linux based operating system comes with all dependencies installed by default. Anyway, check the [INSTALL.txt](https://raw.githubusercontent.com/ucodev/nd-php/master/INSTALL.txt) file for some common exceptions that you might need to do manually.

## Does it only run on Linux?

 The framework was designed to run under any operating system that supports PHP.

## Is it ready for production?

 NO and MAYBE! This is currently a PoC. There's still a lot of work to do such as cleanups, optimizations, refactorings, etc. Use at your own risk and/or unless you really know what your doing.

## Where is the documentation?

 Before reading the documentation, you should see all (or most) of the videos located in the [documentation/videos](https://github.com/ucodev/nd-php/tree/master/documentation/videos) directory.

 The technical documentation of the framework is still being written, but there is a very incomplete version built-in.

 The framework documentation is accessible through any framework web page after installation. Just press the accesskey D (in Firefox is Alt + Shift + D) and the main documentation page will be loaded.

## How to report issues?

 Open the issue on the [ND PHP GitHub Issues Page](https://github.com/ucodev/nd-php/issues). We'll address it as soon as possible.

## Where can I get support?

 As we currently have no community support forum of our own, we encourage you to ask any technical questions on the best Q&A programming community platform available: [StackOverflow](http://stackoverflow.com). The answer may not be that fast as there isn't a lot of people using this framework at the moment, but the question will eventually be answered :).



ChangeLog.txt
v0.6a2 (2018-03-03):

 - Changed datetime fields are now reported back in ISO8601 format for REST JSON update requests.


v0.6a1 (2018-02-03):

 - Fixed field typo on selective deploy condition in the application model.


v0.6a (2018-01-27):

 - Added support for selective deploys (IDE).
 - Added support for unique composite groups (IDE)
 - Added support for field index indication option (IDE).
 - Updated uWeb.
 - Updated language files.


v0.5b3 (2017-10-29):

 - Set headers to lowercase by default.
 - Updated uWeb.
 - Proper server protocol is now set in all responses.
 - Updated tracker.


v0.5b2 (2017-09-17):

 - Fixed conflicting global configuration to local configuration variables.


v0.5b1 (2017-09-17):

 - Corrected expired/non-existent session status codes from Forbidden to Unauthorized.
 - Granted that global configuration setting for sssh custom handlers are enabled before relying on them.


v0.5b (2017-08-20):

 - Fixed minor issues.
 - Minor optimizations.
 - Updated status code list for response module.


v0.5a9 (2017-08-06):

 - Minor fixes on error messages under insert method.
 - Updated uWeb.


v0.5a8 (2017-07-30):

 - Updated tracker.
 - Ignore permission checking on update method for the already evaluated table row perms already.


v0.5a7 (2017-07-30):

 - Some fixes in update procedure.
 - Updated some status codes around.
 - Updated language files.


v0.5a6 (2017-07-30):

 - Minor cleanup.
 - Check compatibility for model deploy based on PHP SAPI information.
 - Updated language files.
 - Updated some status codes around.
 - REST JSON requests for update and insert will now return 403 for unaccessible/non-permitted fields (previous behaviour was to unset fields).


v0.5a5 (2017-07-16):

 - Role assignment checks were too restrictive.
 - Limited typed controllers now allow superuser, admin and superadmin roles to override default values from session properties being assigned to filtering fields.
 - Authentication responses for REST JSON calls now also include the username field in the data structure.
 - Minor cleanup.


v0.5a4 (2017-07-02):

 - Insufficient permissions to search specific fields now cause result requests to explicitly fail (instead of discarding the search).
 - Updated language files.


v0.5a3 (2017-06-29):

 - IDE Builder cleanup.
 - ND Core controller cleanup.
 - Login controller cleanup.


v0.5a2 (2017-06-29):

 - Fixed language file.


v0.5a1 (2017-06-29):

 - Updated tracker.


v0.5a (2017-06-29):

 - Major cleanup and small fixes around.
 - Updated tracker.


v0.04p1 (2017-06-29):

 - Minor improvements all on IDE Builder.


v0.04p (2017-06-29):

 - Base URL settings are no longer retrieved from configuration controller.
 - Cookie Domain segments are now correctly evaluated.
 - IDE Builder now supports Commit Controllers and Apply ACLs operations.
 - Added support for nd-php master/slave nodes.
 - Updated language file.


v0.04o2 (2017-06-25):

 - Fixed value mangling issues with REST JSON listing requests.


v0.04o1 (2017-06-25):

 - Fixed parse error.


v0.04o (2017-06-24):

 - Added default values under registration controller.
 - Updated base model to include four generic string and four generic boolean fields into user model.
 - User role assignments are now allowed only inside users controller.
 - Default authentication mechanism now also supports user email authentication.
 - Enforced admin and superuser role based perm check on object access for limited controller types.
 - Increased default result count limit to 1000 entries for JSON queries.


v0.04n (2017-06-18):

 - Added support for AWS S3 Accelerated Endpoint.
 - Added error logs for failed S3 operations.
 - Improved file upload handlers.
 - Granted that file uploads do not break transactions.
 - Updated uWeb.
 - Enhanced security module.
 - Added config option to disable ND plugins, on a per controller basis.
 - Minor optimizations all around.
 - Optimized caching mechanism.


v0.04m3 (2017-06-02):

 - Minor fixes


v0.04m2 (2017-05-31):

 - Fixed incorrect regex validation return value.


v0.04m1 (2017-05-31):

 - Minor validation improvements.


v0.04m (2017-05-31):

 - Empty multiple relationship fields on view_generic calls are now populated as empty array instead of omitted.
 - Minor fixes around.
 - Session authentication now fully supports userid and apikey based auth.


v0.04l6 (2017-05-30):

 - Proper re-authentication request message is now sent when session expires.
 - Updating _file_ fields to NULL will now cause the file to be removed.
 - Updated language files.


v0.04l5 (2017-05-30):

 - Application model builder now also adds RS perms to users_id field under limited type controllers to non-admin, non-superuser roles.


v0.04l4 (2017-05-29):

 - Check and reject empty arrays on NDSL queries for "in" and "not_in" criteria.


v0.04l3 (2017-05-28):

 - Fixed boolean type check (precedence).


v0.04l2 (2017-05-28):

 - Field types with max length equal to 1 are now considered boolean type.


v0.04l1 (2017-05-28):

 - Boolean types are now correctly converted for JSON REST calls.


v0.04l (2017-05-28):

 - Minor optimizations on Application model.


v0.04k5 (2017-05-28):

 - Improved role permission handling and setup on Application model.
 - Minor improvements on core controllers.
 - Search / Result calls are now fully handled with prepared statements for JSON REST calls.


v0.04k4 (2017-05-27):

 - Updated some status codes all around.
 - Insert and Update operations will now distinguish between NULL, zero and empty fields and accept them as valid values for JSON REST calls.
 - File fields that are NULL, are now ignored by upload pre-processor.


v0.04k3 (2017-05-21):

 - Created and modified times on upload module are now stored in ISO8601 format.


v0.04k2 (2017-05-21):

 - Normalized email validation on both register and login operations.
 - Updated some status codes all around.


v0.04k1 (2017-05-21):

 - Updated uWeb.

user/config/aws.php
<?php if (!defined('FROM_BASE')) { header('HTTP/1.1 403 Forbidden'); die('Invalid requested path.'); }

/* Amazon S3 settings */
$aws['key'] = '';
$aws['secret'] = '';
$aws['version'] = 'latest'; /* '2012-10-17' */
$aws['region'] = 'ap-southeast-1';
$aws['default_bucket'] = 'default';
$aws['bucket_base_dir'] = 'files';
$aws['bucket_img_resize'] = false;
$aws['bucket_img_resize_base_url'] = '';
$aws['bucket_img_resize_xsmall_dir'] = 'xxsmall';
$aws['bucket_img_resize_xsmall_width'] = 60;
$aws['bucket_img_resize_xsmall_dir'] = 'xsmall';
$aws['bucket_img_resize_xsmall_width'] = 120;
$aws['bucket_img_resize_small_dir'] = 'small';
$aws['bucket_img_resize_small_width'] = 240;
$aws['bucket_img_resize_medium_dir'] = 'medium';
$aws['bucket_img_resize_medium_width'] = 480;
$aws['bucket_img_resize_large_dir'] = 'large';
$aws['bucket_img_resize_large_width'] = 960;
$aws['bucket_img_resize_xlarge_dir'] = 'xlarge';
$aws['bucket_img_resize_xlarge_width'] = 1920;
$aws['bucket_img_resize_xxlarge_dir'] = 'xxlarge';
$aws['bucket_img_resize_xxlarge_width'] = 3840;
